gapis/api/vulkan/framegraph.go

```diff
@@ -16,14 +16,153 @@ package vulkan
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/resolve/dependencygraph2"
 	"github.com/google/gapid/gapis/service/path"
 )
 
+// renderpassInfo stores a renderpass' info relevant for the framegraph.
+type renderpassInfo struct {
+	id       uint64
+	beginIdx api.SubCmdIdx
+	endIdx   api.SubCmdIdx
+	nodes    []dependencygraph2.NodeID
+	deps     map[uint64]struct{} // set of renderpasses this renderpass depends on
+}
+
 // GetFramegraph creates the framegraph of the given capture.
 func (API) GetFramegraph(ctx context.Context, p *path.Capture) (*api.Framegraph, error) {
-	log.W(ctx, "TODO: implement framegraph creation")
-	return nil, nil
+	config := dependencygraph2.DependencyGraphConfig{
+		SaveNodeAccesses:    true,
+		ReverseDependencies: true,
+	}
+	dependencyGraph, err := dependencygraph2.GetDependencyGraph(ctx, p, config)
+	if err != nil {
+		return nil, err
+	}
+
+	rpInfos := []*renderpassInfo{}
+	var rpInfo *renderpassInfo
+	currRpId := uint64(0)
+
+	// postSubCmdCb effectively processes each subcommand to extract renderpass info.
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
+		vkState := GetState(state)
+		cmdRef := i.(CommandReferenceʳ)
+		cmdArgs := GetCommandArgs(ctx, cmdRef, vkState)
+
+		// Beginning of renderpass
+		if _, ok := cmdArgs.(VkCmdBeginRenderPassArgsʳ); ok {
+			if rpInfo != nil {
+				panic("Renderpass starts without having ended")
+			}
+			rpInfo = &renderpassInfo{
+				id:       currRpId,
+				beginIdx: subCmdIdx,
+				nodes:    []dependencygraph2.NodeID{},
+				deps:     make(map[uint64]struct{}),
+			}
+			currRpId++
+		}
+
+		// Process commands that are inside a renderpass
+		if rpInfo != nil {
+			nodeID := dependencyGraph.GetCmdNodeID(api.CmdID(subCmdIdx[0]), subCmdIdx[1:])
+			rpInfo.nodes = append(rpInfo.nodes, nodeID)
+		}
+
+		// Ending of renderpass
+		if _, ok := cmdArgs.(VkCmdEndRenderPassArgsʳ); ok {
+			if rpInfo == nil {
+				panic("Renderpass ends without having started")
+			}
+			rpInfo.endIdx = subCmdIdx
+			rpInfos = append(rpInfos, rpInfo)
+			rpInfo = nil
+		}
+	}
+
+	// Iterate on the capture commands to gather information
+	c, err := capture.ResolveGraphicsFromPath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	if err := sync.MutateWithSubcommands(ctx, p, c.Commands, nil, nil, postSubCmdCb); err != nil {
+		return nil, err
+	}
+
+	updateDependencies(rpInfos, dependencyGraph)
+
+	// Build the framegraph nodes and edges from collected data.
+	nodes := make([]*api.FramegraphNode, len(rpInfos))
+	for i, rpInfo := range rpInfos {
+		// Graphviz DOT: use "\l" as a newline to obtain left-aligned text.
+		text := fmt.Sprintf("Renderpass %v\\lbegin:%v\\lend:%v\\l", rpInfo.id, rpInfo.beginIdx, rpInfo.endIdx)
+		nodes[i] = &api.FramegraphNode{
+			Id:   rpInfo.id,
+			Text: text,
+		}
+	}
+
+	edges := []*api.FramegraphEdge{}
+	for _, rpInfo = range rpInfos {
+		for deps := range rpInfo.deps {
```

Comment:
A notifier: This is a map and in Golang this will iterate differently for every range operation for the same data. If this is considered there is no problem.

Reply:
Yep, no worries there.